### PR TITLE
ENH: Increase average calculation performance. See #5507 

### PR DESCRIPTION
**Pull request by minhlongdo**

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -514,8 +514,7 @@ def average(a, axis=None, weights=None, returned=False):
         avg = a.mean(axis)
         scl = avg.dtype.type(a.size/avg.size)
     else:
-        a = a + 0.0
-        wgt = np.asarray(weights)
+        wgt = np.asarray(weights, dtype=float)
```
Comment (Nodd) [inline numpy/lib/function_base.py]: As said in the coresponding issue, you should use ```np.result_type(weights, np.float)``` instead of ```float``` only.

Comment (juliantaylor) [inline numpy/lib/function_base.py]: unfortunately that won't work as result_type needs an ndarray as input, which is was asarray creates

Comment (njsmith) [inline numpy/lib/function_base.py]: Maybe asarray needs an argument to say 'make the dtype at least this high
on the casting graph (higher is OK)'?

**Comment by njsmith** [comment]
On further consideration, this change doesn't make much sense to me. The
speedup is from dropping the a = a + 0.0 -- how does changing the dtype of
weights replace this?

OTOH assuming that weights is real is probably OK -- I thought we were
forcing 'a' to be real.

BTW, copy=0 is the default for asarray.

**Comment by jaimefrio** [comment]
I think the best solution would look something like:

```
wts = np.asarray(wts)
dt = np.result_type(a.dtype, wts.dtype, float)
```

and later on always make calls to the ufuncs with `dtype=dt` specified:

```
scl = wgt.sum(axis=axis, dtype=dt)
avg = np.multiply(a, wgt, dtype=dt).sum(axis) / scl
```

There is also a very weird construct a little bit further down:

```
if returned:
    scl = np.multiply(avg, 0) + scl
```

which we could probably do away with if we implemented the above, as the only conceivable reason I can see to do that would be to have `scl` be of `np.result_type(avg.dtype, scl.dtype)`.

**Comment by ewmoore** [comment]
Doesn't `np.result_type(a.dtype, wts.dtype, float)` upcast float32 to
float64?

**Comment by seberg** [comment]
Yes, you would need to use `np.result_type(a, wts, 0.)` (or use `a.dtype` if you want it to never behave like a scalar for `a`, etc.)

File: numpy/lib/function_base.py (hunk 1 above)
Comment (seberg) [inline numpy/lib/function_base.py]: Guys, the calculation down below already does a result type for weights, you do not need to cast it here (unless you want at least 64bit floats). As for a, you can actually use result type to cast it as suggested, and just do the cast in the sum like it is done for the weights?

**Comment by minhlongdo** [comment]
In the current code base by doing <code> a = a + 0.0 </code> is to force a to be real. So I was wondering if <code> np.average(a, weights=None, returned=False) </code> also accepts an array of complex numbers. Because, if it accepts <code> np.complex128 </code> then the operation <code> a = a + 0.0 </code> would not be necessary since the imaginary and real parts of the complex number are already <code> np.float64 </code>. So I did the following in the if statement <code> if not isinstance(a, np.matrix) </code> to reduce unnecessary operations that forces <code> a </code> to be real.
<code>
```
if a.dtype is np.dtype('complex128'): 
    if weights is None:
        # Just convert into an np.array
        a = np.asarray(a)
    else:
        if a.dtype is not np.dtype('complex128'):
            # Convert to np.array and set np.dtype to float64
            a = np.asarray(a, dtype=np.dtype('float64'))
else:
    # If a is a np.matrix and there are weight values
    if weights is not None:
        # if dtype of a is not np.dtype('complex128') then make it to be np.float64
        if a.dtype is not np.dtype('complex128'):
            # No need to attempt conversion for np.float64 if it is already np.float64
            if a.dtype is not np.dtype('float64'):
                # Worse case scenario, np.matrix is not np.float64 then the following will force a to be np.float64
                a = a + 0.0
```
 </code>

The performance of this change by using the script https://gist.github.com/skuschel/2d148a37a2ce17925fb0 provided by <code>skuschel</code> is the following:

execution time np.average: 3.06e-01 sec
exectution time np.mean: 3.00e-01 sec

What do you guys think?

**Comment by minhlongdo** [comment]
Sorry I accidentally closed it.

**Comment by charris** [comment]
>In the current code base by doing `a = a + 0.0` is to force a to be real

Real or complex while preserving inexact type. It is an tricky way to achieve that end. Maybe we should just check for `inexact`? This PR  looks a bit too complicated to me as it stands.

@minhlongdo At some point the commits will need to be squashed into one and the commit message cleaned up. See `doc/source/dev/gitwash/development_workflow.rst` for discussion of commit messages.

**Comment by minhlongdo** [comment]
@charris Will do thanks for the feedback.

**Comment by minhlongdo** [comment]
@charris I looked into <code>inexact</code> and it does indeed simply the proposed solution.
I somehow managed to mess up, so I made a new pull request https://github.com/numpy/numpy/pull/5551. I am sorry for the mess.

**Pull request closed without merging**